### PR TITLE
Add refresh_events to COSAgentProvider to update relation data

### PR DESCRIPTION
**Pull request by chanchiwai-ray**
Add `refresh_events` to `COSAgentProvider` to update relation data when

- config changed (default behavior)
- upgrade charm.

### CODE REVIEW

**Comment by chanchiwai-ray** [comment]
We could do some manual tests about it. However, the unit test should already be able check the behavior.

File: src/charm.py [hunk 1]
```diff
@@ -33,6 +33,7 @@ def __init__(self, *args: Any) -> None:
 
         self.cos_agent_provider = COSAgentProvider(
             self,
+            refresh_events=[self.on.config_changed, self.on.upgrade_charm],
```
Comment (jneo8) [inline src/charm.py:36]: **praise (non-blocking):** 

Look nice. Can we add some comment here to explain the behavior?

**Review by jneo8 — APPROVED** [review]
LGTM.

One non-blocking comment.

File: src/charm.py (hunk 1 above)
Comment (chanchiwai-ray) [reply src/charm.py:39]: DONE

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -268,3 +269,75 @@ def test_13_update_status_config_invalid(self, mock_hw_tool_helper, mock_exporte
...
+        self.harness.charm.on.update_status.emit()
+        self.assertEqual(self.harness.charm.unit.status, ActiveStatus("Unit is ready"))
+
+        relation_data = self.harness.get_relation_data(rid, "hardware-observer/0")
+        metrics_alert_rules = json.loads(relation_data["config"]).get("metrics_alert_rules")
+
+        with mock.patch(
+            "charm.COSAgentProvider._metrics_alert_rules", new_callable=mock.PropertyMock
+        ) as mock_alert_rules:
+            fake_metrics_alert_rules = {}  # just make it empty
```
Comment (dashmage) [inline tests/unit/test_charm.py:298]: nit: maybe we don't need a comment here?

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -268,3 +269,75 @@ def test_13_update_status_config_invalid(self, mock_hw_tool_helper, mock_exporte
...
```
Comment (dashmage) [inline tests/unit/test_charm.py:334]: nit: same here

File: tests/unit/test_charm.py [hunk 4]
```diff
@@ -268,3 +269,75 @@ def test_13_update_status_config_invalid(self, mock_hw_tool_helper, mock_exporte
...
+        relation_data = self.harness.get_relation_data(rid, "hardware-observer/0")
+        updated_metrics_alert_rules = json.loads(relation_data["config"]).get(
+            "metrics_alert_rules"
+        )
+        self.assertEqual(updated_metrics_alert_rules, fake_metrics_alert_rules)
+        self.assertNotEqual(updated_metrics_alert_rules, metrics_alert_rules)
+
+    @mock.patch("charm.Exporter", return_value=mock.MagicMock())
+    @mock.patch("charm.HWToolHelper", return_value=mock.MagicMock())
+    def test_15_upgrade_charm_update__alert_rules(self, mock_hw_tool_helper, mock_exporter):
```
Comment (dashmage) [inline tests/unit/test_charm.py:311]: typo: Remove extra `_` between update and alert

**Review by dashmage — APPROVED** [review]
Left some small comments. But mostly LGTM :smile:

File: tests/unit/test_charm.py (hunk 4 above)
Comment (chanchiwai-ray) [reply tests/unit/test_charm.py]: Done:+1:

**Pull request merged**